File: domain-driven-design/csharp/src/DDD.Domain/Customers/Address.cs
namespace DDD.Domain.Customers;

using DDD.Domain.Common;

public class Address : ValueObject
{
    public string Street { get; }
    public string City { get; }
    public string State { get; }
    public string Country { get; }
    public string ZipCode { get; }
    
    private Address(string street, string city, string state, string country, string zipCode)
    {
        Street = street;
        City = city;
        State = state;
        Country = country;
        ZipCode = zipCode;
    }
    
    public static Address Create(string street, string city, string state, string country, string zipCode)
    {
        if (string.IsNullOrWhiteSpace(street))
            throw new ArgumentException("Street cannot be empty", nameof(street));
            
        if (string.IsNullOrWhiteSpace(city))
            throw new ArgumentException("City cannot be empty", nameof(city));
            
        if (string.IsNullOrWhiteSpace(state))
            throw new ArgumentException("State cannot be empty", nameof(state));
            
        if (string.IsNullOrWhiteSpace(country))
            throw new ArgumentException("Country cannot be empty", nameof(country));
            
        if (string.IsNullOrWhiteSpace(zipCode))
            throw new ArgumentException("ZipCode cannot be empty", nameof(zipCode));
            
        return new Address(street, city, state, country, zipCode);
    }
    
    protected override IEnumerable<object> GetEqualityComponents()
    {
        yield return Street;
        yield return City;
        yield return State;
        yield return Country;
        yield return ZipCode;
    }
    
    public override string ToString()
    {
        return $"{Street}, {City}, {State}, {Country} {ZipCode}";
    }
}

File: domain-driven-design/java/src/com/example/domain/model/valueobject/Address.java
package com.example.domain.model.valueobject;

import java.util.Objects;

/**
 * 住所を表す値オブジェクト
 */
public final class Address {
    private final String streetAddress;
    private final String city;
    private final String state;
    private final String zipCode;
    private final String country;

    private Address(String streetAddress, String city, String state, String zipCode, String country) {
        this.streetAddress = streetAddress;
        this.city = city;
        this.state = state;
        this.zipCode = zipCode;
        this.country = country;
    }

    public static Address of(String streetAddress, String city, String state, String zipCode, String country) {
        Objects.requireNonNull(streetAddress, "Street address cannot be null");
        Objects.requireNonNull(city, "City cannot be null");
        Objects.requireNonNull(state, "State cannot be null");
        Objects.requireNonNull(zipCode, "Zip code cannot be null");
        Objects.requireNonNull(country, "Country cannot be null");

        if (streetAddress.isBlank()) {
            throw new IllegalArgumentException("Street address cannot be empty");
        }
        if (city.isBlank()) {
            throw new IllegalArgumentException("City cannot be empty");
        }
        if (state.isBlank()) {
            throw new IllegalArgumentException("State cannot be empty");
        }
        if (zipCode.isBlank()) {
            throw new IllegalArgumentException("Zip code cannot be empty");
        }
        if (country.isBlank()) {
            throw new IllegalArgumentException("Country cannot be empty");
        }

        return new Address(streetAddress, city, state, zipCode, country);
    }

    // 値オブジェクトは変更のためのメソッドを持つ場合、新しいインスタンスを返す
    public Address withStreetAddress(String streetAddress) {
        return new Address(streetAddress, this.city, this.state, this.zipCode, this.country);
    }

    public Address withCity(String city) {
        return new Address(this.streetAddress, city, this.state, this.zipCode, this.country);
    }

    public Address withState(String state) {
        return new Address(this.streetAddress, this.city, state, this.zipCode, this.country);
    }

    public Address withZipCode(String zipCode) {
        return new Address(this.streetAddress, this.city, this.state, zipCode, this.country);
    }

    public Address withCountry(String country) {
        return new Address(this.streetAddress, this.city, this.state, this.zipCode, country);
    }

    // 値オブジェクトのドメインロジック
    public boolean isSameCountryAs(Address other) {
        return this.country.equalsIgnoreCase(other.country);
    }

    public boolean isInternational(String baseCountry) {
        return !this.country.equalsIgnoreCase(baseCountry);
    }

    // ゲッター
    public String getStreetAddress() {
        return streetAddress;
    }

    public String getCity() {
        return city;
    }

    public String getState() {
        return state;
    }

    public String getZipCode() {
        return zipCode;
    }

    public String getCountry() {
        return country;
    }

    // 値オブジェクトの等価性は、全てのプロパティの値が等しいことで判断
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Address address = (Address) o;
        return Objects.equals(streetAddress, address.streetAddress) &&
                Objects.equals(city, address.city) &&
                Objects.equals(state, address.state) &&
                Objects.equals(zipCode, address.zipCode) &&
                Objects.equals(country, address.country);
    }

    @Override
    public int hashCode() {
        return Objects.hash(streetAddress, city, state, zipCode, country);
    }

    @Override
    public String toString() {
        return streetAddress + ", " + city + ", " + state + " " + zipCode + ", " + country;
    }

    // フォーマットされた住所を返す
    public String getFormattedAddress() {
        return String.format("%s\n%s, %s %s\n%s", 
                streetAddress, city, state, zipCode, country);
    }
}